gsdk/codegen/src/storage.rs

```diff
@@ -0,0 +1,131 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2021-2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::{quote, ToTokens};
+use syn::{
+    parse_quote, punctuated::Punctuated, Expr, ExprLit, FnArg, Ident, ItemFn, Lit, LitStr, Meta,
+    MetaNameValue, PatType,
+};
+
+const FULL_DOC_SUFFIX: &str = " at specified block";
+const FULL_SUFFIX: &str = "_at";
+
+/// Storage query builder for generating
+/// - basic private shared query
+/// - full query with block hash
+/// - short query with block hash None.
+pub struct StorageQueryBuilder(ItemFn);
+
+impl StorageQueryBuilder {
+    /// Build storage query from full blocks.
+    fn full(&self) -> ItemFn {
+        let mut short = self.0.clone();
+
+        // reset function docs.
+        //
+        // - remove `at specified block` suffix.
+        short.attrs.iter_mut().find_map(|attr| {
+            if let Meta::NameValue(MetaNameValue {
+                value:
+                    Expr::Lit(ExprLit {
+                        attrs: _,
+                        lit: Lit::Str(lit_str),
+                    }),
+                ..
+            }) = &mut attr.meta
+            {
+                *lit_str = LitStr::new(
+                    &lit_str.value().replace(FULL_DOC_SUFFIX, ""),
+                    lit_str.span(),
+                );
+                return Some(());
+            }
+
+            None
+        });
+
+        // reset function name.
+        //
+        // - `storage_query_at` -> `storage_query`
+        short.sig.ident = Ident::new(
+            short.sig.ident.to_string().trim_end_matches(FULL_SUFFIX),
```

Comment:
Don't we want to add some validation here for ident that it end with `_at` ? Same for comment and args